flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/AbstractResourceManagerDriver.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.resourcemanager.active;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.clusterframework.types.ResourceIDRetrievable;
+import org.apache.flink.runtime.concurrent.ScheduledExecutor;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract common base class for implementations of {@link ResourceManagerDriver}.
+ */
+public abstract class AbstractResourceManagerDriver<WorkerType extends ResourceIDRetrievable>
+	implements ResourceManagerDriver<WorkerType> {
+
+	protected final Logger log = LoggerFactory.getLogger(getClass());
+
+	protected final Configuration flinkConfig;
+	protected final Configuration flinkClientConfig;
```

Comment:
The background is that, RM will not send the complete configuration to the TMs, but only the differences compared to the original configuration submitted by the client.

Currently, when submitting the job, client will ship the local `flink-conf.yaml` file. This file is available to both JM and TMs, through Yarn Shared Cache and Kubernetes ConfigMap.

TM should not use that original `flink-conf.yaml` directly, because the configuration might be overwritten on the JM side (e.g., dynamic properties, JM address, TM resource, etc.). Therefore, RM will compare the effective configuration (`flinkConfig`) with the original one shipped from the client side (`flinkClientConfig`), and send the differences to TMs as dynamic properties, which overwrites the original configuration on the TM side.

Reply:
I think in the general case the assumption is wrong that `flinkClientConfig` is the original configuration because we don't load the configuration file from a potentially configured configuration directory. It might hold true wrt to the current start up scripts but it is brittle since it ties very specific assumptions together.

Moreover, where is the difference between sending the full configuration vs. a diff which is applied to the presumably same base version? Shouldn't it result in the same effective configuration? Of course we should send the full `taskManagerConfig` and not the JM configuration.

Reply:
Hi @tillrohrmann thanks for pointing out this concern. 

> Why we are trying to only pass the diff config options, rather that the full configuration?

That will make the start commands very long. Also it is hard to distinguish the config options only for TaskManager. So currently, we compare the effective configuration in ResourceManager and client shipped configuration. And then pass the differences.

> Is `flinkClientConfig` the original configuration?

I think yes. For Yarn and K8s entrypoints(e.g. `YarnJobClusterEntrypoint`, `KubernetesApplicationClusterEntrypoint`, etc.), we do not accept the configured configuration directory(via `--configDir`). Instead we load them from a specific directory(`$PWD` for Yarn and `ENV_FLINK_CONF_DIR` for K8s), both of them are guaranteed correctly setting in the Flink client.

However, if we could find out which config options should be passed the TaskManagers, i agree that `flinkClientConfig` could be removed.

Reply:
It might be ok at the moment because the current entrypoints ensure that the assumptions are respected. However, nobody says that this won't change in the future. The farther these assumptions are away from each other (e.g. `KubernetesResourceManagerDriver` assumes that the corresponding entrypoint sets up the right env variables), the more brittle it is. Hence, I would suggest to rethink the usage of `GlobalConfiguration` in the future.

Reply:
Yeah, i agree with you to rethink the usage of `GlobalConfiguration` and how to ship the dynamic config options to TaskManagers.